---
title: "Course Intro"
subtitle: "Practical Computing Skills for Omics Data (PLNTPTH 5004)" 
author: Jelmer Poelstra
date: 2025-08-26
institute: "MCIC Wooster, Ohio State University"
editor: visual
title-slide-attributes:
  data-background-color: "dimgray"
format:
  revealjs:
    theme: [simple, ../theme_mod.scss]
    smaller: true
    scrollable: false
    progress: true
    incremental: false
    css: ../slides.css
    slide-number: true
    slide-level: 2
    highlight-style: pygments
    embed-resources: true
    toc: true
    toc-depth: 1
    toc-title: Overview
editor_options: 
  chunk_output_type: console
---

# Personal introductions {background-color="dimgray"}

## Introductions: Jelmer

-   Lead of Bioinformatics and Microscopy at MCIC in Wooster since 2020
    -   The MCIC is a CFAES core facility that provides services in the areas of
        molecular biology, high-throughput sequencing, bioinformatics, and microscopy.
    -   The majority of my time is spent providing research assistance,\
        working with grad students and postdocs on mostly genomic & transcriptomic data
    -   I also teach: some courses, workshops, Code Club (<https://osu-codeclub.github.io>)

. . .

<hr style="height:1pt; visibility:hidden;" />

-   Background in animal evolutionary genomics & speciation

<hr style="height:1pt; visibility:hidden;" />

-   In my free time, I enjoy bird watching -- locally & all across the world

## Introductions: You

-   Name

-   Lab and Department

-   Research interests and/or current research topics

-   Something about you that is *not* work-related

# Course overview {background-color="dimgray"}

## The core goals of this course

Learning skills to:

-   Do your research more reproducibly and efficiently

-   Prepare yourself for working with large "omics" datasets

## Course background: Reproducibility

-   Two related ideas:

    1.  Getting same results with an independent experiment (*replicable*)

    2.  Getting same results given the same data (*reproducible*)

    Our focus is on #2.

------------------------------------------------------------------------

## Course background: Reproducibility (cont.)

> *"The most basic principle for reproducible research is: Do everything via code."*\
> —Karl Broman

<br>

-   Also important:

    -   Project organization and documentation (*week 2*)

    -   Sharing data and code (*for code: Git & GitHub, week 3*)

    -   How you code (*e.g. week 4 - shell scripts, and 6 - Nextflow*)

<br>

::: callout-tip
##### Another motivator: working reproducibly will benefit future you!
:::

## Course background: Efficiency and automation

-   Using code enables you to work more efficiently and automatically —\
    particularly useful when having to:

    -   Do repetitive tasks

    -   Recreate a figure or redo an analysis after adding a sample

    -   Redo a project after uncovering a mistake in the first data processing step.

## Course background: Omics data

-   Omics data is increasingly important in biology, and most notably includes:
    -   Genomics, including Metagenomics
    -   Transcriptomics
    -   Proteomics
    -   Metabolomics

. . .

<hr style="height:1pt; visibility:hidden;" />

-   While we'll be using some example omics datasets, this course will *not* teach you how to analyze omics data in full --- our focus is on fundamental computational skills.

# Course topics {background-color="dimgray"}

## The Unix shell & shell scripts (Wk 2-4 and more)

Being able to work in the Unix shell is a fundamental skill in computational biology.

<hr style="height:1pt; visibility:hidden;" />

-   You'll spend a lot of time with the Unix shell, starting this week, and including in weeks that aren't solely focused on the shell.
-   We'll also write shell *scripts*, and will use an editor called VS Code for this and other purposes.

<br>

::::::: columns
:::: column
![](img/bash.png){fig-align="center" width="50%"}

::: legend2
Bash (shell language)
:::
::::

:::: column
![](img/vscode.png){fig-align="center" width="50%"}

::: legend2
VS Code
:::
::::
:::::::

## Project organization & documentation (Wk 3 & 6) {.nostretch}

Good project file organization & documentation is a necessary starting point for reproducible research.

<hr style="height:1pt; visibility:hidden;" />

-   You'll learn best practices for project organization, file naming, etc.

-   To document and report what you are doing, you'll use *Markdown* files.

-   Data and software management (TBA)

<br>

![](img/markdown.png){fig-align="center" width="35%"}

::: legend2
Markdown
:::

## Version control with Git and GitHub (Wk 4)

Using version control, you can more effectively keep track of project progress,
collaborate, share code, revisit earlier versions, and undo.

<hr style="height:1pt; visibility:hidden;" />

-   *Git* is the version control software we will use,\
    and *GitHub* is the website that hosts Git projects (repositories).

-   You'll also use Git+GitHub to hand in your graded assignments.

<br>

::::: columns
::: {.column width="50%"}
![](img/git.png){fig-align="center" width="65%"}
:::

::: {.column width="50%"}
![](img/github.jpg){fig-align="center"}
:::
:::::

## High-performance computing with OSC (Wk 6-7)

Thanks to supercomputer resources,
you can work with **very large datasets at speed** ---
running up to 100s of analyses in parallel,
and using much larger amounts of memory and storage space than a personal computer has.

<hr style="height:1pt; visibility:hidden;" />

. . .

-   We will use OSC throughout the course, and you'll get a brief intro to it today.
-   In week 5, you'll learn how to manage data and software at OSC.
-   In week 6, we'll learn to submit shell scripts as OSC "batch jobs" with Slurm.

<br>

:::::: columns
::: {.column width="40%"}
![](img/osc.png){fig-align="center" width="100%"}
:::

::: {.column width="28%"}
![](img/slurm.png){fig-align="center" width="75%"}
:::

::: {.column width="31%"}
![](img/conda.png){fig-align="center" width="95%"}
:::
::::::

## Automated workflow management (Wk 8-11) {.nostretch}

Using a workflow written with a workflow manager,
you can run and rerun entire analysis pipeline with a single command,
and easily change and rerun parts of it, too.

<hr style="height:1pt; visibility:hidden;" />

-   We'll use the workflow language *Nextflow*.
-   MORE DETAILS TBA

## R (Wk 12-15) {.nostretch}

-   DETAILS TBA

<br>

![](img/nextflow.png){fig-align="center" width="50%"}

# Course practicalities {background-color="dimgray"}

## Zoom

-   Be muted by default, but feel free to unmute yourself to ask questions any time.

-   Questions can also be asked in the chat.

-   Having your camera turned on as much as possible is appreciated!

. . .

-   "Screen real estate" — large/multiple monitors or multiple devices best.

-   Be ready to share your screen.

## Websites & Books

-   Info about CarmenCanvas website...

<hr style="height:1pt; visibility:hidden;" />

-   The GitHub website contains:
    -   Overviews of each week & readings
    -   Slide decks and lecture pages
    -   Exercises
    -   Final project assignment information

<hr style="height:1pt; visibility:hidden;" />

. . .

-   Books:

    -   Computing Skills for Biologists ("CSB"; Allesina & Wilmes 2019)
    -   Bioinformatics Data Skills ("Buffalo"; Buffalo 2015)

# Homework and grading {background-color="dimgray"}

## What your grade is made up of

You can earn a total of 100 points across 6 assignments and 4 final project
checkpoints.

## Graded assignments

These are due on Mondays and are worth 10 points each:

1. Shell basics (week 3)
2. Markdown & Git (week 5)
3. Shell scripting (week 6)
4. OSC batch jobs (week 8)
5. Nextflow (week 11)
6. R (week 14)

## Final project

Plan and implement a small computational project, with the following checkpoints:

-   I: **Proposal** (due week 13 -- 5 points)

-   II: **Draft** (due week 15 -- 5 points)

-   III: **Oral presentations** on Zoom (week 16 -- 10 points)

-   IV: **Final submission** (due Dec 15 -- 20 points)

<hr style="height:1pt; visibility:hidden;" />

. . .

::: callout-tip
#### Data sets for the final project

If you have your own data set & analysis ideas, that is ideal.
If not, I can provide you with this.

More information about the final project will follow later in the course.
:::

## Ungraded homework

-   Weekly readings — somewhat up to you when to do these, ideally before *and* after the lectures!

-   Weekly exercises — I recommend doing these on Fridays

-   Miscellaneous small assignments such as surveys and account setup.

<br>

::: callout-note
### Weekly materials & homework

I will try add the materials for each week on the preceding Friday ---
at the least the week's overview and readings.

None of this homework had to be handed in.
:::

## Weekly recitation on Monday

We will have an optional weekly recitation meeting on Monday in
which we go through the exercises for the preceding week.

<br>

If you're interested, indicate your availability here: TBA

## Rest of this week

-   Introduction to omics data

<hr style="height:1pt; visibility:hidden;" />

-   Brief introduction to the Ohio Supercomputer Center (OSC)

-   **Homework**:

    - TBA

# Questions? {background-color="dimgray"}

<br><br><br><br>

[(Back to the site)](/index.qmd)
