---
title: "Week 1: Intro to the course, omics data, and OSC"
pagetitle: "PRACS25: Week 1 overview"
author: Jelmer Poelstra
date: 2025-08-25
---

---------

<br>

## Content overview

This week, you will get an overview of the course,
and introductions to omics data and the Ohio Supercomputer Center (OSC).
More specifically, you will learn:

#### [Course intro](w1_01_course-intro.qmd) (Tuesday class)

- What to expect from this course
- Which tools and languages we will use
- What is expected of you during this course
- Get up to speed on the infrastructure of the course

#### [Omics data](w1_02_omics-data.qmd) (Tuesday/Thursday class)

- What omics data is
- About the main high-throughput sequencing technologies
- How Illumina sequence libraries are structured
- What reference genomes are and are good for
- What the main sequence file types are and what FASTA and FASTQ files look like

#### [OSC Intro](w1_03_osc.qmd) (Thursday class)

- What a supercomputer is and why they are useful
- What resources the Ohio Supercomputer Center (OSC) provides
- How to access OSC resources through its OnDemand webportal

<hr style="height:1pt; visibility:hidden;" />

## Assignments and exercises

#### Ungraded assignments

- Pre-course survey - link TBA
- [Ohio Supercomputer Center account setup](w1_ua_osc-setup.qmd) [_(This must be done by Thursday's class)_]{style="color:#bb0000"}

_No exercises this week._

<hr style="height:1pt; visibility:hidden;" />

## Readings

#### Required readings

- [Poinsignon et al. 2023](https://link-springer-com.proxy.lib.ohio-state.edu/protocol/10.1007/978-1-0716-3195-9_10) _Working with Omics Data: An Interdisciplinary Challenge at the Crossroads of Biology and Computer Science_
- [Lee 2023](https://www.ksdb.org/archive/view_article?pid=dr-27-1-9) _The Principles and Applications of High-Throughput Sequencing Technologies_

#### Optional readings

**Book chapters**

Throughout the course, chapters from these two books are optional readings for this course.
Both can be accessed online for free with OSU credentials via the links below.

- [Bioinformatics Data Skills (Buffalo 2015)](https://library.ohio-state.edu/record=b9490023~S7) ("**Buffalo**" for short)\
  This week: Preface and Chapter 1 - How _to Learn Bioinformatics_

- [Computing Skills for Biologists (Allesina & Wilmes 2019)](https://library.ohio-state.edu/record=b8624007~S7) ("**CSB**" for short)\
  This week: Chapter 0 - _Building a Computing Toolbox_ (Introduction & Section 0.1 only)

**Papers**

What a list ðŸ˜³! Well, the idea is that you can browse through these titles and
see if any of these spark your interest ---
many of these present omics and HTS data in a more field-specific light, e.g. plant pathology.

- [von der Heyden et al. 2025](https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.14123) _Advancing species conservation and management through omics tools_
- [Nizamani et al. 2023](https://phytopatholres.biomedcentral.com/articles/10.1186/s42483-023-00199-5) _High-throughput sequencing in plant disease management: a comprehensive review of benefits, challenges, and future perspectives_
- [Aragona et al. 2022](https://pmc.ncbi.nlm.nih.gov/articles/PMC9320658/) _New-generation sequencing technology in diagnosis of fungal plant pathogens: A dream comes true?_
- [Mahmood et al. 2022](https://www.frontiersin.org/journals/plant-science/articles/10.3389/fpls.2022.1062952/full) _Multi-omics revolution to promote plant breeding efficiency_
- [Rai et al. 2019](https://www.sciencedirect.com/science/article/abs/pii/S2452310019300216) _A new era in plant functional genomics_ TBA OSU LINK
- [Pinto & Bhatt 2024](https://www.nature.com/articles/s41576-024-00746-6) _Sequencing-based analysis of microbiomes_
- [Manzoni et al. 2018](https://academic.oup.com/bib/article/19/2/286/2562648) _Genome, transcriptome and proteome: the rise of omics data and their integration in biomedical sciences_
- [van Dijk et al. 2023](https://www-cell-com.proxy.lib.ohio-state.edu/trends/genetics/fulltext/S0168-9525(23)00119-1) _Genomics in the long-read sequencing era_
